pkg/controllers/signer.go

```diff
@@ -0,0 +1,205 @@
+
+type GoogleCAS struct {
+	client client.Client
+
+	MaxRetryDuration time.Duration
+}
+
+func (s *GoogleCAS) SetupWithManager(ctx context.Context, mgr ctrl.Manager) error {
+	const fieldOwner = "cas-issuer.jetstack.io"
+
+	if err := cmapi.AddToScheme(mgr.GetScheme()); err != nil {
+		return err
+	}
+
+	if err := issuersv1beta1.AddToScheme(mgr.GetScheme()); err != nil {
+		return err
+	}
+
+	controller := &controllers.CombinedController{
+		IssuerTypes:        []issuerapi.Issuer{&issuersv1beta1.GoogleCASIssuer{}},
+		ClusterIssuerTypes: []issuerapi.Issuer{&issuersv1beta1.GoogleCASClusterIssuer{}},
+
+		FieldOwner:       fieldOwner,
+		MaxRetryDuration: s.MaxRetryDuration,
+
+		Sign:  s.Sign,
+		Check: s.Check,
+
+		EventRecorder: mgr.GetEventRecorderFor(fieldOwner),
+	}
+
+	return controller.SetupWithManager(ctx, mgr)
+}
+
+func (o *GoogleCAS) extractIssuerSpec(obj client.Object) (issuerSpec *issuersv1beta1.GoogleCASIssuerSpec, namespace string) {
+	switch t := obj.(type) {
+	case *issuersv1beta1.GoogleCASIssuer:
+		return &t.Spec, t.Namespace
+	case *issuersv1beta1.GoogleCASClusterIssuer:
+		return &t.Spec, viper.GetString("cluster-resource-namespace")
+	}
+
+	panic("Program Error: Unhandled issuer type")
+}
+
+func (o *GoogleCAS) Check(ctx context.Context, issuerObj client.Object) error {
+	issuerSpec, resourceNamespace := o.extractIssuerSpec(issuerObj)
+
+	casClient, _, err := o.createCasClient(ctx, resourceNamespace, issuerSpec)
+	if err != nil {
+		return err
+	}
+	casClient.Close()
+
+	return nil
+}
+
+// Sign implements signer.Sign for Venafi TPP and Venafi-as-a-Service.
+func (o *GoogleCAS) Sign(ctx context.Context, cr *cmapi.CertificateRequest, issuerObj client.Object) ([]byte, error) {
+	issuerSpec, resourceNamespace := o.extractIssuerSpec(issuerObj)
+
+	casClient, parent, err := o.createCasClient(ctx, resourceNamespace, issuerSpec)
+	if err != nil {
+		return nil, signer.IssuerError{Err: err}
+	}
+	defer casClient.Close()
+
+	duration := apiutil.DefaultCertDuration(cr.Spec.Duration)
+	if duration < cmapi.MinimumCertificateDuration {
+		duration = cmapi.MinimumCertificateDuration
+	}
+
+	createCertificateRequest := &casapi.CreateCertificateRequest{
+		Parent: parent,
+		// Should this use the certificate request name?
+		CertificateId: fmt.Sprintf("cert-manager-%d", rand.Int()),
+		Certificate: &casapi.Certificate{
+			CertificateConfig: &casapi.Certificate_PemCsr{
+				PemCsr: string(cr.Spec.Request),
+			},
+			Lifetime: &durationpb.Duration{
+				Seconds: duration.Milliseconds() / 1000,
+				Nanos:   0,
+			},
+		},
+		RequestId:                     uuid.New().String(),
+		IssuingCertificateAuthorityId: issuerSpec.CertificateAuthorityId,
+	}
+
+	createCertResp, err := casClient.CreateCertificate(ctx, createCertificateRequest)
+	if err != nil {
+		return nil, fmt.Errorf("casClient.CreateCertificate failed: %w", err)
+	}
+
+	// TODO: for backwards compatibility, we should also set the CA field in the certificaterequest
```

Comment:
This is currently breaking, we have to update issuer-lib to support adding the CA in the certificaterequest.